src/nisarqa/processing/insar/quiver_plots.py

```diff
@@ -352,26 +492,70 @@ def truncate_colormap(
     arrow_stride = int(max(np.shape(total_offset)) / params.arrow_density)
 
     # Only plot the arrows at the requested strides.
-    arrow_y = az_off[::arrow_stride, ::arrow_stride]
-    arrow_x = rg_off[::arrow_stride, ::arrow_stride]
+    az_offsets_at_arrow_tails = az_off[::arrow_stride, ::arrow_stride]
+    rg_offsets_at_arrow_tails = rg_off[::arrow_stride, ::arrow_stride]
 
-    x = np.linspace(0, arrow_x.shape[1] - 1, arrow_x.shape[1])
-    y = np.linspace(0, arrow_x.shape[0] - 1, arrow_x.shape[0])
+    num_arrows_along_x = az_offsets_at_arrow_tails.shape[1]
+    num_arrows_along_y = az_offsets_at_arrow_tails.shape[0]
+    x = np.linspace(0, num_arrows_along_x - 1, num_arrows_along_x)
+    y = np.linspace(0, num_arrows_along_y - 1, num_arrows_along_y)
     X, Y = np.meshgrid(x, y)
 
+    # 2D grid of the starting x pixel indices in the image arrays for each arrow
+    pixel_indices_of_arrow_tails_in_x_direction = X * arrow_stride
+
+    # 2D grid of the starting y pixel indices in the image arrays for each arrow
+    pixel_indices_of_arrow_tails_in_y_direction = Y * arrow_stride
+
+    # Determine the offsets in the image grid's X and Y directions
+
+    if quiver_projection_params is None:
+        # ROFF, RIFG, RUNW
+        # For L1 (ROFF, etc), `az_offsets_at_arrow_tails` indicates how many
+        # meters to offset in the image grid's y-direction (aka along-track).
+        # (Similar reasoning for slant range offsets and the x-direction.)
+        y_offsets_at_arrow_tails = az_offsets_at_arrow_tails
+        x_offsets_at_arrow_tails = rg_offsets_at_arrow_tails
+    else:
+        # GOFF, GUNW
+        # For L2 (GOFF, etc), `az_offsets_at_arrow_tails` indicates how many
+        # meters to offset in the satellite's along-track direction, which
+        # is NOT the same as the image grid's y-direction. So, do conversion.
+        if (x_coordinates is None) or (y_coordinates is None):
+            raise ValueError(
+                f"`quiver_projection_params` was not None, but"
+                f" {x_coordinates=} and {y_coordinates=} (neither can be None)."
+            )
+
+        x_offsets_at_arrow_tails, y_offsets_at_arrow_tails = (
+            get_offset_values_in_projected_coordinates(
+                az_offsets=az_offsets_at_arrow_tails,
+                rg_offsets=rg_offsets_at_arrow_tails,
+                x_coordinates=x_coordinates[::arrow_stride],
+                y_coordinates=y_coordinates[::arrow_stride],
+                projection_params=quiver_projection_params,
+            )
+        )
+
+        # Invert y component because matplotlib.pyplot.quiver follows the
+        # axes convention in the background image (positive y).
+        # But, the offset images follow the top-down convention (negative y),
+        # so we need to invert the values.
```

Comment:
```suggestion
        # NOTE: Invert the Y component because `matplotlib.pyplot.quiver()` after `matplotlib.pyplot.imshow()` will follow the coordinate convention of the raster i.e. row index downward. This is opposite to the coordinate system in which the offset was defined (i.e. +y upward), so the y offset needs to be inverted to properly plot the quivers on top of the image.
```

Please feel free to edit and split the lines.

Reply:
Oh wow, this is a great explanation! Thank you!
(And thank you for figuring out the bug, too!)